### ISSUE TITLE
SmartTemplates ignored when creating an email from the Thunderbird taskbar context menu

### ISSUE THREAD

**Issue by goozak** [RealRaven2000/SmartTemplates#272]
Windows 10 / Thunderbird 115.4.1 / SmartTemplates 4.3.2

When right-clicking the Thunderbird icon on the taskbar and selecting "Write new message", the message created doesn't contain the template usually applied to new message (like when I use the "+ New Message" button above the folder list).

**Comment by RealRaven2000** [comment]
Hmm, it works for me, can you export your template for the account like this:

![image](https://github.com/RealRaven2000/SmartTemplates/assets/6314674/f970e4ab-e3c1-4a86-a1b9-f2cd016b75b2)

If you like you can send it to me directly / off Github, maybe mention `[SmartTemplates] issue 272` in the subject line.

Note that it is triggered by your email address/ identity  as long as the option _[x] Apply the following template to the new message_ is activated.

**Comment by RealRaven2000** [comment]
ok, first of all I tried to import your settings this morning, and to my surprise it didn't load them into my test profile. So there could be a separate regression in Tb115 that breaks the import, I will investigate that first. See issue #273 - I will fix this quickly and then import your settings to test them.

**Comment by RealRaven2000** [comment]
After repairing the import, I am loading your template but it seems to work for me, here is the composer after loading (tested both without and with ThunderHTML Edit plugin)

![image](https://github.com/RealRaven2000/SmartTemplates/assets/6314674/ff83c36d-c595-43cb-bf48-7c491c246099)

It generates the following HTML in composer:

```html
<!DOCTYPE html>
<html>
  <head>

    <meta http-equiv="content-type" content="text/html; charset=">
  </head>
  <body smarttemplateinserted="true">
    <div id="smartTemplate4-template">
      <div style="font-family:'Trebuchet MS',sans-serif;font-size:10pt">
        <p> </p>
        <br>
      </div>
    </div>
    <p><br>
    </p>
  </body>
</html>

```


In _Plaintext mode_ it generates this markup:

```html
<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=">
    <style></style>
  </head>
  <body style="font-family: -moz-fixed; white-space: pre-wrap; width: 72ch;" smarttemplateinserted="true"><div
  id="smartTemplate4-template"><div style="font-family:'Trebuchet MS',sans-serif;font-size:10pt"><p> </p><br></div></div></body>
</html>
```

but then... I realize I haven't read your description correctly, I was using this interface:

![image](https://github.com/RealRaven2000/SmartTemplates/assets/6314674/2bb8138a-f3b8-4f82-a349-2278d3631ad1)

You wrote something about the (Windows) taskbar... I have actually never done this, so let me look into this way of creating an email.

**Comment by RealRaven2000** [comment]
So the first thing I notice here is that the "`notifyComposeBodyReady`" event is not fired when Thunderbird opens composer in the way you have described. this one is quite crucial for SmartTemplates to work, but it's probably not the only possible entry point for triggering its logic. You said this was working in older versions of Thunderbird (102) ?

**Comment by goozak** [comment]
> You said this was working in older versions of Thunderbird (102) ?

Not sure about that, to be honest.  I don't use that  context menu option often.

**Comment by RealRaven2000** [comment]
OK, I found the culprit. 

Apparently the window is missing some parameters that I needed to determine composeType. If it is a _reply / forward_ composeType, I try to extract all headers from the original mail. All this data doesn't exist and the composer startup logic throws an error so SmartTemplates doesn't get to execute any of its logic. 

I am not sure whether there are other ways to trigger this logic (maybe if you click on a `mailto:` link on a web page, this I need to investigate separately. But I think this will also trigger the "new mail" compose type, never the reply or forward. So I guess I have dealt with this problem now, by falling back to the global `window.gComposeType` which will tell me that it is a New email and I do not need to extract any previous headers.

[smartTemplate-fx-4.3.3pre7.zip](https://github.com/RealRaven2000/SmartTemplates/files/13249367/smartTemplate-fx-4.3.3pre7.zip)

---

to install the test version, download the zip file and then drag the file into Thunderbird Add-ons Manager (do not extract contents).

PS: I will close the issue myself once the next version including bugfix is released.

**Comment by goozak** [comment]
4.3.3pre7 works as expected.  :-)

_Interestingly, 'dragging into Add-ons Manager' doesn't work if the Add-ons Manager is opened via the "Add-ons and Themes" button that can be added to the Mail Toolbar...  Works as expected when going thru the menu button._

**Comment by RealRaven2000** [comment]
> 4.3.3pre7 works as expected. :-)
> 
> _Interestingly, 'dragging into Add-ons Manager' doesn't work if the Add-ons Manager is opened via the "Add-ons and Themes" button that can be added to the Mail Toolbar... Works as expected when going thru the menu button._

Hmm, that  worked for me - but I could only add the button to the settings tabs - I couldn't actually find the button for the "mail" space.

**Comment by goozak** [comment]
> I couldn't actually find the button for the "mail" space.

For me, that button is in the "Available for all Spaces" section and is on all the toolbars...  Weird.

**Comment by RealRaven2000** [comment]
I have now also tested the case when a "mailto" link is clicked on the website - same problem here. This version fixes that by also checking the `composeType` for being `Ci.nsIMsgCompType.MailToUrl`.

[smartTemplate-fx-4.3.3pre11.zip](https://github.com/RealRaven2000/SmartTemplates/files/13260025/smartTemplate-fx-4.3.3pre11.zip)

-----

to install the test version, download the zip file and then drag the file into Thunderbird Add-ons Manager (do not extract contents).

**Comment by RealRaven2000** [comment]
Fixed in v 4.3.3 released Nov 8th.

**Comment by RealRaven2000** [comment]
See also version 4.9 - published 11/11/2024
